sdk/src/main/scala/com/horizen/account/state/StateTransition.scala

```diff
@@ -0,0 +1,118 @@
+package com.horizen.account.state
+
+import com.horizen.account.utils.BigIntegerUtil
+import scorex.util.ScorexLogging
+
+import java.math.BigInteger
+
+class StateTransition(
+    view: AccountStateView,
+    messageProcessors: Seq[MessageProcessor],
+    blockGasPool: GasPool,
+    blockContext: BlockContext
+) extends ScorexLogging {
+
+  @throws(classOf[InvalidMessageException])
+  @throws(classOf[ExecutionFailedException])
+  def transition(msg: Message): Array[Byte] = {
+    // do preliminary checks
+    preCheck(msg)
+    // allocate gas for processing this message
+    val gasPool = buyGas(msg)
+    // consume intrinsic gas
+    val intrinsicGas = GasUtil.intrinsicGas(msg.getData, msg.getTo == null)
+    if (gasPool.getGas.compareTo(intrinsicGas) < 0) throw IntrinsicGasException(gasPool.getGas, intrinsicGas)
+    gasPool.subGas(intrinsicGas)
+    // find and execute the first matching processor
+    messageProcessors.find(_.canProcess(msg, view)) match {
+      case None =>
+        log.error(s"No message processor found for executing message $msg")
+        throw new IllegalArgumentException("Unable to process message.")
+      case Some(processor) =>
+        // increase the nonce by 1
+        view.increaseNonce(msg.getFrom.address())
+        // create a snapshot to rollback to in case of execution errors
+        val revisionId = view.snapshot
+        try {
+          processor.process(msg, view, gasPool, blockContext)
+        } catch {
+          // if the processor throws ExecutionRevertedException we revert all changes
+          case err: ExecutionRevertedException =>
+            view.revertToSnapshot(revisionId)
+            throw err
+          // if the processor throws ExecutionFailedException we revert all changes and consume any remaining gas
+          case err: ExecutionFailedException =>
+            view.revertToSnapshot(revisionId)
+            gasPool.subGas(gasPool.getGas)
+            throw err
+          // any other exception will bubble up and invalidate the block
+        } finally {
+          refundGas(msg, gasPool)
```

Comment:
do we need to refund gas in all the cases?
For success and `ExecutionRevertedException` seems ok.
But what about  `ExecutionFailedException`? I think, we want to consume all gas.
@JcBernack pls, check in geth